web/docs/auth/migrate-from-11.md

```diff
@@ -0,0 +1,318 @@
+  const users = await prismaClient.user.findMany({
+    include: {
+      auth: true,
+    },
+  });
+
+  for (const user of users) {
+    if (user.auth) {
+      console.log("User was already migrated, skipping", user);
+      continue;
+    }
+
+    const providerData: UsernameProviderData = {
+      hashedPassword: user.password,
+    };
+    const providerName: ProviderName = "username";
+
+    await prismaClient.auth.create({
+      data: {
+        identities: {
+          create: {
+            providerName,
+            providerUserId: user.username.toLowerCase(),
+            providerData: JSON.stringify(providerData),
+          },
+        },
+        user: {
+          connect: {
+            id: user.id,
+          },
+        },
+      },
+    });
+  }
+}
+```
+
+
+### Email
+
+```ts title="src/server/migrateToNewAuth.ts"
+import { PrismaClient } from "@prisma/client";
+import { EmailProviderData, ProviderName } from "@wasp/auth/utils";
+
+export async function migrateEmailAuth(prismaClient: PrismaClient) {
+  const users = await prismaClient.user.findMany({
+    include: {
+      auth: true,
+    },
+  });
+
+  for (const user of users) {
+    if (user.auth) {
+      console.log("User was already migrated, skipping", user);
+      continue;
+    }
+
+    if (!user.email || !user.password) {
+      console.log("Missing email auth info, skipping user", user);
+      continue;
+    }
+
+    const providerData: EmailProviderData = {
+      isEmailVerified: user.isEmailVerified,
+      emailVerificationSentAt:
+        user.emailVerificationSentAt?.toISOString() ?? null,
+      passwordResetSentAt: user.passwordResetSentAt?.toISOString() ?? null,
+      hashedPassword: user.password,
+    };
+    const providerName: ProviderName = "email";
+
+    await prismaClient.auth.create({
+      data: {
+        identities: {
+          create: {
+            providerName,
+            providerUserId: user.email,
+            providerData: JSON.stringify(providerData),
+          },
+        },
+        user: {
+          connect: {
+            id: user.id,
+          },
+        },
+      },
+    });
+  }
+}
+```
+
+
+### Google & GitHub
+
+```ts title="src/server/migrateToNewAuth.ts"
+import { PrismaClient } from "@prisma/client";
+import { ProviderName } from "@wasp/auth/utils";
+
+export async function migrateGoogleAuth(prismaClient: PrismaClient) {
+  return createSocialLoginMigration(prismaClient, "google");
+}
+
+export async function migrateGitHubAuth(prismaClient: PrismaClient) {
+  return createSocialLoginMigration(prismaClient, "github");
+}
+
+async function createSocialLoginMigration(
+  prismaClient: PrismaClient,
+  providerName: "google" | "github"
+) {
+  const users = await prismaClient.user.findMany({
+    include: {
+      auth: true,
+      externalAuthAssociations: true,
+    },
+  });
+
+  for (const user of users) {
+    if (user.auth) {
+      console.log("User was already migrated, skipping", user);
```

Comment:
Is it possible that user was migrated but with another identity? No, right, because right now they can have only one type of identity? However what if they first ran "username" script and that one migrated the user but the user also has oauth creds and now it will skip it here? But that is not possible because we don't allow one user to have both user/pass and oauth, is that so?

Reply:
One auth identity per user. Which ever they migrate first will take the identity. This is explained in the Auth Entities docs a bit, but they shouldn't face these issues since mostly they used either Google or email or username.

Reply:
But they could have used both, right? It was allowed. Should we warn them then about this, so they can choose how to handle this? This is connected with my comment above, where I said that since we have this regression, we should probably let them know about it, before they start doing migration, so they can prepare for it and decide how they are going to handle that.